### PR TITLE
Dom manipulation to sort icon cards

### LINKED ISSUE #86: Tabbing order confusion
### Kind of issue 

- [ ] Feature
- [ ] Bug
- [ ] Design
- [x] Other, namely: a11y

### Description

Currently, tabbing through the website really only works if you're using the default ordering (alphabetical). If either colored ordering or relevance ordering (when searching) is used, tabbing through the grid items will result in jumping all over the page.

This is because the visual re-ordering of the grid items is achieved using the flexbox `order` property and dynamically assigned CSS variables. The idea behind this approach, when it was implemented, was to leverage the GPU to do the re-ordering.

Unfortunately, this is a minor (?) problem for accessibility and if performance isn't too bad when shuffling the elements around to preserve a sane tab ordering we should change the implementation.

### PR DESCRIPTION
**Pull request by jorgeamadosoria**
according to #86 , CSS ordering is to be substituted with JS sorting for colors to allow tabbing to work.
This PR includes DOM sorting for all criteria, related refactorings and unit and e2e tests for this new approach.
DOM rearranging is slower than the previous solution, but that was expected and is the only solution for the tabbing problem (without updating tabIndex everywhere).

### CODE REVIEW

**Review by ericcornelissen — COMMENTED** [review]
Please resolve the merge conflicts and undo unwanted changes (e.g. `package*.json`) first :slightly_smiling_face:

**Comment by mondeja** [comment]
Would you mind to solve the conflicts @jorgeamadosoria?

**Comment by jorgeamadosoria** [comment]
> Would you mind to solve the conflicts @jorgeamadosoria?

conflicts are solved.

**Comment by mondeja** [comment]
New conflict in `public/scripts/index.js` 🔧

**Review by mondeja — COMMENTED** [review]
New conflict in `public/scripts/index.js` 🔧

**Comment by jorgeamadosoria** [comment]
> New conflict in `public/scripts/index.js` 🔧

this is done.

File: public/stylesheet.css [hunk 1]
```diff
@@ -282,6 +284,29 @@ a {
   padding: 0.2rem;
 }
 
+/* Extensions */
```
Comment (mondeja) [inline public/stylesheet.css]: Why this is being added again? Please, rebase properly.

**Review by mondeja — DISMISSED** [review]

File: public/index.pug [hunk 2]
```diff
@@ -37,12 +37,12 @@ html(lang='en-US')
...
           )
 
-  body.no-js.order-alphabetically.download-svg(tabindex='-1')
+  body.no-js.order-alpha.download-svg(tabindex='-1')
     .popup_modal.popup__top.hidden
       .popup-body 
         table.extensions__table
           tr
-            td Extension
+            td Extensions
```
Comment (mondeja) [inline public/index.pug]: This renaming has no sense. The column of the table should be singular.

File: public/index.pug [hunk 3]
```diff
@@ -304,10 +307,10 @@ html(lang='en-US')
           src='//cdn.carbonads.com/carbon.js?serve=CKYIPK7M&placement=simpleiconsorg',
           type='text/javascript'
         )
-
         each icon in icons
           li.grid-item(
-            style=`--order-color: ${icon.indexByColor}`,
+            order-color=`${icon.indexByColor}`,
+            order-alpha=`${icon.indexByAlpha}`,
             data-brand=`${icon.normalizedName}`
```
Comment (mondeja) [inline public/index.pug]: These values of attributes can be defined in pug using the variables directly:

```suggestion
            order-color=icon.indexByColor,
            order-alpha=icon.indexByAlpha,
            data-brand=icon.normalizedName
```

File: tests/ordering.test.js [hunk 4]
```diff
@@ -121,30 +122,30 @@ describe('Ordering', () => {
...
 
-    initOrdering(document, localStorage);
+    initOrdering(document, localStorage, domUtils);
     expect(localStorage.hasItem).toHaveBeenCalledWith(STORAGE_KEY_ORDERING);
     expect(localStorage.getItem).toHaveBeenCalledWith(STORAGE_KEY_ORDERING);
-    expect(localStorage.setItem).not.toHaveBeenCalled();
+    expect(localStorage.setItem).toHaveBeenCalledTimes(1);
   });
 
   it('uses the stored value "color"', () => {
```
Comment (mondeja) [inline tests/ordering.test.js]:
```suggestion
  it('uses the stored value "order-color"', () => {
```

File: public/stylesheet.css [hunk 5]
```diff
@@ -536,7 +538,7 @@ input[type='search']::-ms-reveal {
 }
 
 /* Control - order */
-body.order-alphabetically #order-alpha,
+body.order-alpha #order-alpha,
 body.order-by-color #order-color,
 body.order-by-relevance #order-relevance {
```
Comment (mondeja) [inline public/stylesheet.css]: These selectors must be updated as well.

```suggestion
body.order-color #order-color,
body.order-relevance #order-relevance {
```

File: tests/dom-utils.test.js [hunk 6]
```diff
@@ -2,14 +2,33 @@ const {
...
 } = require('../public/scripts/dom-utils.js');
 
+const orderedData = [
+  { 'order-alpha': 1, getAttribute: (order) => 1 },
+  { 'order-alpha': 2, getAttribute: (order) => 2 },
+  { 'order-alpha': 3, getAttribute: (order) => 3 },
+];
+const testData = [
+  { 'order-alpha': 3, getAttribute: (order) => 3 },
+  { 'order-alpha': 2, getAttribute: (order) => 2 },
```
Comment (mondeja) [inline tests/dom-utils.test.js]: To avoid a false negative in the test if the implementation reverse ordering you could invert the order of these two items:

```suggestion
  { 'order-alpha': 2, getAttribute: (order) => 2 },
  { 'order-alpha': 3, getAttribute: (order) => 3 },
```

**Review by mondeja — DISMISSED** [review]

File: public/index.pug (hunk 2 above)
Comment (jorgeamadosoria) [reply public/index.pug]: as you wish.

File: public/scripts/dom-utils.js [hunk 7]
```diff
@@ -19,6 +19,16 @@ export function toggleVisibleElement($el) {
...
 
+export function sortChildren($el, attribute) {
+  [...$el.children]
+    .sort(
+      (a, b) =>
+        parseInt(a.getAttribute(attribute)) -
+        parseInt(b.getAttribute(attribute)),
+    )
+    .forEach((node) => $el.appendChild(node));
+}
```
Comment (mondeja) [inline public/scripts/dom-utils.js:30]: Here is a improved async version of this function to do the ordering instantly:

```suggestion
  const sorted = [...$el.children]
    .sort(
      (a, b) =>
      parseInt(a.getAttribute(attribute)) -
      parseInt(b.getAttribute(attribute)),
    )
  $el.innerHTML = ""
  sorted.slice(0, 15).forEach((node) => $el.appendChild(node))
  setTimeout(async () => {
    sorted.slice(15).forEach((node) => $el.appendChild(node));
  }, 5);
```

Resetting the HTML content and appending synchronously only the first 15 icons do the trick. I suppose that this number can be increased without decreasing performance.

File: public/scripts/ordering.js [hunk 8]
```diff
@@ -1,44 +1,42 @@
...
   let preferredOrdering = DEFAULT_ORDERING;
 
   const $body = document.querySelector('body');
-  const $orderAlphabetically = document.getElementById('order-alpha');
-  const $orderByColor = document.getElementById('order-color');
-  const $orderByRelevance = document.getElementById('order-relevance');
+  const $orderAlpha = document.getElementById('order-alpha');
+  const $orderColor = document.getElementById('order-color');
+  const $orderRelevance = document.getElementById('order-relevance');
+  const $gridIcons = document.querySelector('ul.grid');
```
Comment (mondeja) [inline public/scripts/ordering.js]: This variable is not used:

```suggestion
```

**Review by mondeja — CHANGES_REQUESTED** [review]

File: public/scripts/dom-utils.js (hunk 7 above)
Comment (mondeja) [inline public/scripts/dom-utils.js]: Here is a improved async version of this function to do the ordering "instantly". Really doesn't does the ordering instantly, but seems like it because the icons placed at the top are reordered and added first to the grid:

```suggestion
export async function sortChildren($el, attribute) {
  const sorted = [...$el.children]
    .sort(
      (a, b) =>
      parseInt(a.getAttribute(attribute)) -
      parseInt(b.getAttribute(attribute)),
    )
  const firstChildren = sorted.slice(0, 15);
  $el.innerHTML = "";
  firstChildren.forEach((node) => $el.appendChild(node));
  setTimeout(async () => {
    sorted.slice(15).forEach((node) => $el.appendChild(node));
  }, 1);
}
```

Resetting the HTML content and appending first only the first 15 icons do the trick. I suppose that this number can be increased without decreasing performance.

**Review by mondeja — CHANGES_REQUESTED** [review]

File: public/scripts/ordering.js [hunk 9]
```diff
@@ -1,44 +1,42 @@
...
-  $orderByColor.disabled = false;
-  $orderByRelevance.disabled = false;
+  $orderAlpha.disabled = false;
+  $orderColor.disabled = false;
+  $orderRelevance.disabled = false;
 
   if (storage.hasItem(STORAGE_KEY_ORDERING)) {
     const storedOrdering = storage.getItem(STORAGE_KEY_ORDERING);
     selectOrdering(storedOrdering);
+    domUtils.sortChildren(document.querySelector('ul.grid'), storedOrdering);
```
Comment (mondeja) [inline public/scripts/ordering.js]: This line is not required, the ordering is executed two times in this case:

```suggestion
```

**Review by mondeja — CHANGES_REQUESTED** [review]

File: public/scripts/dom-utils.js (hunk 7 above)
Comment (jorgeamadosoria) [reply public/scripts/dom-utils.js]: Isn't setTimeout going to create a new anon function per 15 elements, thus eating memory?
What happens if you sort after having scrolled down a bit (thus the elements you see are not the first 15)?

Comment (mondeja) [reply public/scripts/dom-utils.js]: > Isn't setTimeout going to create a new anon function per 15 elements, thus eating memory?

Not sure if I follow you, `slice(15)` means from 15 to the end. `setTimeout` is executed one time. 

> What happens if you sort after having scrolled down a bit (thus the elements you see are not the first 15)?

That is the problem. The page scrolls to the top or the bottom when the ordering is changed. Not sure if could be worth to store the scroll height pixel before sorting and reset the same scroll position after. Something like:

```javascript
const scrollY = window.scrollY;
// after ordering
window.scrollTo(0, scrollY);
```

This can be improved computing what cards should be added first depending on the position of the screen.

**Comment by jorgeamadosoria** [comment]
I can't answer on the previous thread, don't know why. 

I thought setTimeout was making more than one call. If it is like you said, then nevermind, this is fine.

I have a different issue:

This is also not going to be enough for the compact layout. Maybe instead of 15 the slice should be 30 or even 50 (compact can have up to 9 cards per line, so 3 lines would be 27 instead of 15).
Maybe that also takes care of most of the cases if people have done a little bit of scrolling, not too much.

More importantly: visually, there is not that big of a delay between ordering the whole list and ordering it by slices, so I don't know if the gain is enough to justify the complexity. This is just my opinion, however: if it is up to me, simpler is better than faster, unless the code is too slow for confort. I don't think the sorting is too slow for confort just yet.

In short, this is clever, but I don't think there is any need for it. If you still think we need it, I would recommend a larger slice to account for compact (30 cards should be fine), a comment along the lines of "Above the Fold item sorting" or something like that on the first call, and "rendering the rest of the sorted items" in the setTimeout. Something to mitigate the complexity.

Other than that, the code works just fine.

**Comment by mondeja** [comment]
I think that the async implementation is necessary because the blocking would be a great performance regression.

After reflecting on it a bit, seems that the scroll before sorting should be setted to 0 in axis Y, so the page would be seen at the start. Currently, the scroll behaves erratically when sorting because travels to some random point of the page, so it is an improvement anyways.

```javascript
    $body.classList.add(selected);
    window.scrollTo(0, 0);
    domUtils.sortChildren(document.querySelector('ul.grid'), selected);
```

With this line and the asynchronous implementation seems to me that it should be fine, unless I'm missing something happening when sorting by relevance on search.

**Comment by jorgeamadosoria** [comment]
> I think that the async implementation is necessary because the blocking would be a great performance regression.
> 
> After reflecting on it a bit, seems that the scroll before sorting should be setted to 0 in axis Y, so the page would be seen at the start. Currently, the scroll behaves erratically when sorting because travels to some random point of the page, so it is an improvement anyways.
> 
> ```js
>     $body.classList.add(selected);
>     window.scrollTo(0, 0);
>     domUtils.sortChildren(document.querySelector('ul.grid'), selected);
> ```
> 
> With this line and the asynchronous implementation seems to me that it should be fine, unless I'm missing something happening when sorting by relevance on search.

Scrolling to the beginning is definitely good, I agree.
I would make a domUtils function for setting the scrolling view value. Or you can leave it like that and I will abstract that later on. 

If you are going to keep this, then merge it at will.

**Comment by mondeja** [comment]
Some improvements in the tests and this will be ready 👍🏼

**Review by mondeja — APPROVED** [review]
LGTM
